src/Services/TagsNormalizerService.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace App\Services;
+
+use App\Entity\Tag;
+use App\Repository\TagRepository;
+use Doctrine\ORM\EntityManagerInterface;
+
+final class TagsNormalizerService implements TagsNormalizerInterface
+{
+    public function __construct(
+        private readonly EntityManagerInterface $em,
+        private readonly TagRepository $tagRepository,
+    ) {
+    }
+
+    /**
+     * Normalize list of names to ensure they fit the DB schema.
+     *
+     * @param array $names
+     *   Array with tags names
+     *
+     * @return array array<strings>
+     *   Normalized array with tag names
+     */
+    public function normalize(array $names): array
+    {
+        if (!empty($names)) {
+            $names = $this->trimLength($names);
+            $names = $this->normalizeToDbName($names);
+        }
+
+        return $names;
+    }
+
+    /**
+     * Trim names in list to ensure they fit the DB schema.
+     *
+     * @param array $names
+     *   Array with tags names
+     *
+     * @return array
+     *   Normalized array with tag names
+     */
+    private function trimLength(array $names): array
+    {
+        $metadata = $this->em->getClassMetadata(Tag::class);
+        $maxNameLength = isset($metadata->fieldMappings, $metadata->fieldMappings['name'], $metadata->fieldMappings['name']['length']) ? (int) $metadata->fieldMappings['name']['length'] : 50;
+
+        // Ensure we don't exceed field length in db
+        return array_map(function ($name) use ($maxNameLength) {
+            return mb_substr(trim($name), 0, $maxNameLength);
+        }, $names);
+    }
+
+    /**
+     * Normalize to the database tag name.
+     *
+     * @todo: Make it possible to limit normalization against single controlled vocabulary.
+     *
+     * @param array $names
+     *   Array with tags names
+     *
+     * @return array
+     *   Normalized array with tag names
+     */
+    private function normalizeToDbName(array $names): array
+    {
+        $normalizedNames = [];
+        foreach ($names as $name) {
+            $tag = $this->tagRepository->findOneBy(['name' => $name]);
```

Comment:
It's not explicitly stated, but I assume you're normalizing casing of tag names here.

Performing a database query per tag name is not a very efficient solution. Using something like

```php
$tags = $this->tagRepository->findBy(['name' => $names]);
$tagNames = array_map(static fn ($tag) => $tag->getName(), $tags);
…         
```
must be better. It may even be worth bypassing the entity manager and perform a raw SQL query.

Reply:
Not sure it's that important in this async flow. But I don't see how you would known which tags have been found and normalized? How would you `$tagNames` and match them back to which `$names` they matched and which they did not?

Think this would make it some-what more complex.